File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/ListView/LVYearly.java
package wxm.KeepAccount.ui.data.show.note.ListView;

import android.content.Context;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ListView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.OnClick;
import wxm.KeepAccount.ui.base.Adapter.LVAdapter;
import wxm.KeepAccount.utility.ToolUtil;
import wxm.androidutil.ViewHolder.ViewDataHolder;
import wxm.androidutil.ViewHolder.ViewHolder;
import wxm.androidutil.util.UtilFun;
import wxm.KeepAccount.R;
import wxm.KeepAccount.ui.base.Helper.ResourceHelper;
import wxm.KeepAccount.ui.data.show.note.ACNoteShow;
import wxm.KeepAccount.ui.data.show.note.ShowData.FilterShowEvent;
import wxm.KeepAccount.ui.data.show.note.base.ValueShow;
import wxm.KeepAccount.ui.utility.HelperDayNotesInfo;
import wxm.KeepAccount.ui.utility.ListViewHelper;
import wxm.KeepAccount.ui.utility.NoteDataHelper;
import wxm.KeepAccount.ui.utility.NoteShowInfo;
import wxm.KeepAccount.utility.ContextUtil;
import wxm.uilib.IconButton.IconButton;

/**
 * 年数据视图辅助类
 * Created by WangXM on 2016/9/10.
 */
public class LVYearly extends LVBase {
    private final LinkedList<String> mLLSubFilter = new LinkedList<>();
    private final LinkedList<View> mLLSubFilterVW = new LinkedList<>();

    // 若为true则数据以时间降序排列
    private boolean mBTimeDownOrder = true;
    private boolean mBSelectSubFilter = false;

    // for main listview
    class MainAdapterItem   {
        public String  tag;
        public String  show;

        public String  year;
        public recordDetail yearDetail;
        public String  amount;

        MainAdapterItem()    {
            show = EShowFold.FOLD.getName();
            yearDetail = new recordDetail();
        }
    }

    class MainItemHolder extends ViewDataHolder<String, MainAdapterItem> {
        public MainItemHolder(String tag)   {
            super(tag);
        }

        @Override
        protected MainAdapterItem getDataByTag(String tag) {
            NoteShowInfo ni = NoteDataHelper.Companion.getInfoByYear(tag);

            MainAdapterItem map = new MainAdapterItem();
            map.year = tag;
            map.yearDetail.mPayCount = String.valueOf(ni.getPayCount());
            map.yearDetail.mIncomeCount = String.valueOf(ni.getIncomeCount());
            map.yearDetail.mPayAmount = String.format(Locale.CHINA,
                    "%.02f", ni.getPayAmount());
            map.yearDetail.mIncomeAmount = String.format(Locale.CHINA,
                    "%.02f", ni.getIncomeAmount());

            BigDecimal bd_l = ni.getBalance();
            map.amount = String.format(Locale.CHINA,
                    (0 < bd_l.floatValue()) ? "+ %.02f" : "%.02f", bd_l);

            map.tag = tag;
            map.show = EShowFold.getByFold(!checkUnfoldItem(tag)).getName();
            return map;
        }
    }
    protected final LinkedList<MainItemHolder> mMainPara;

    // for sub listview
    class SubAdapterItem   {
        public String  tag;
        public String  subTag;

        public String  month;
        public recordDetail monthDetail;
        public String  amount;

        SubAdapterItem()    {
            monthDetail = new recordDetail();
        }
    }

    class SubItemHolder extends ViewDataHolder<String, SubAdapterItem>    {
        public SubItemHolder(String tag)   {
            super(tag);
        }

        @Override
        protected SubAdapterItem getDataByTag(String tag) {
            String ky = tag.substring(0, 4);
            NoteShowInfo ni = NoteDataHelper.Companion.getInfoByMonth(tag);
            SubAdapterItem map = new SubAdapterItem();

            String km = tag.substring(5, 7);
            km = km.startsWith("0") ? km.replaceFirst("0", " ") : km;
            map.month = km;
            map.monthDetail.mPayCount = String.valueOf(ni.getPayCount());
            map.monthDetail.mIncomeCount = String.valueOf(ni.getIncomeCount());
            map.monthDetail.mPayAmount = String.format(Locale.CHINA,
                    "%.02f", ni.getPayAmount());
            map.monthDetail.mIncomeAmount = String.format(Locale.CHINA,
                    "%.02f", ni.getIncomeAmount());

            BigDecimal bd_l = ni.getBalance();
            map.amount = String.format(Locale.CHINA,
                    (0 < bd_l.floatValue()) ? "+ %.02f" : "%.02f", bd_l);

            map.tag = ky;
            map.subTag = tag;
            return map;
        }
    }
    protected final HashMap<String, LinkedList<SubItemHolder>> mHMSubPara;


    class YearlyActionHelper extends ActionHelper    {
        @BindView(R.id.ib_sort)
        IconButton mIBSort;

        @BindView(R.id.ib_report)
        IconButton mIBReport;

        @BindView(R.id.ib_delete)
        IconButton mIBDelete;

        @BindView(R.id.ib_add)
        IconButton mIBAdd;

        YearlyActionHelper() {
            super();
        }

        @Override
        protected void initActs() {
            mIBReport.setVisibility(View.GONE);
            mIBAdd.setVisibility(View.GONE);
            mIBDelete.setVisibility(View.GONE);

            mIBSort.setActIcon(mBTimeDownOrder ?
                    R.drawable.ic_sort_up_1 : R.drawable.ic_sort_down_1);
            mIBSort.setActName(mBTimeDownOrder ?
                    R.string.cn_sort_up_by_time : R.string.cn_sort_down_by_time);
        }

        @OnClick({R.id.ib_sort, R.id.ib_refresh})
        public void onActionClick(View v) {
            switch (v.getId()) {
                case R.id.ib_sort: {
                    mBTimeDownOrder = !mBTimeDownOrder;

                    mIBSort.setActIcon(mBTimeDownOrder ? R.drawable.ic_sort_up_1 : R.drawable.ic_sort_down_1);
                    mIBSort.setActName(mBTimeDownOrder ? R.string.cn_sort_up_by_time : R.string.cn_sort_down_by_time);

                    reorderData();
                    loadUI(null);
                }
                break;

                case R.id.ib_refresh: {
                    reloadView(v.getContext(), false);
                }
                break;
            }
        }
    }


    public LVYearly() {
        super();
        mBActionExpand = false;

        mMainPara = new LinkedList<>();
        mHMSubPara = new HashMap<>();
        mAHActs = new YearlyActionHelper();
    }

    @Override
    protected boolean isUseEventBus() {
        return true;
    }

    /**
     * 过滤视图事件
     *
     * @param event 事件
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onFilterShowEvent(FilterShowEvent event) {
    }



    /**
     * "接受"或者"取消"后动作
     *
     * @param v 动作view
     */
    @OnClick({R.id.bt_accpet, R.id.bt_giveup})
    public void onAccpetOrGiveupClick(View v) {
        int vid = v.getId();
        switch (vid) {
            case R.id.bt_accpet:
                if (mBSelectSubFilter) {
                    if (!UtilFun.ListIsNullOrEmpty(mLLSubFilter)) {
                        ACNoteShow ac = getRootActivity();
                        ac.jumpByTabName(NoteDataHelper.TAB_TITLE_MONTHLY);

                        ArrayList<String> al_s = new ArrayList<>(mLLSubFilter);
                        EventBus.getDefault().post(new FilterShowEvent(NoteDataHelper.TAB_TITLE_YEARLY, al_s));

                        mLLSubFilter.clear();
                    }

                    for (View i : mLLSubFilterVW) {
                        i.setSelected(false);
                        i.getBackground().setAlpha(0);
                    }
                    mLLSubFilterVW.clear();

                    mBSelectSubFilter = false;
                    refreshAttachLayout();
                }
                break;

            case R.id.bt_giveup:
                mBSelectSubFilter = false;
                mLLSubFilter.clear();

                for (View i : mLLSubFilterVW) {
                    i.setSelected(false);
                    i.getBackground().setAlpha(0);
                }
                mLLSubFilterVW.clear();

                refreshAttachLayout();
                break;
        }
    }

    /**
     * 重新加载数据
     */
    @Override
    protected void initUI(Bundle bundle) {
        super.initUI(bundle);

        ToolUtil.INSTANCE.runInBackground(getActivity(),
                () -> {
                    mMainPara.clear();
                    mHMSubPara.clear();

                    // for year
                    List<String> set_k = NoteDataHelper.Companion.getNotesYears();
                    Collections.sort(set_k, (o1, o2) -> !mBTimeDownOrder ? o1.compareTo(o2) : o2.compareTo(o1));
                    for(String k : set_k) {
                        mMainPara.add(new MainItemHolder(k));
                    }

                    // for month
                    List<String> set_k_m = NoteDataHelper.Companion.getNotesMonths();
                    Collections.sort(set_k_m, (o1, o2) -> !mBTimeDownOrder ? o1.compareTo(o2) : o2.compareTo(o1));
                    for(String k : set_k_m) {
                        String ky = k.substring(0, 4);
                        LinkedList<SubItemHolder> lsDay = mHMSubPara.get(ky);
                        if(null == lsDay)   {
                            lsDay = new LinkedList<>();
                            mHMSubPara.put(ky, lsDay);
                        }
                        lsDay.add(new SubItemHolder(k));
                    }
                },
                () -> loadUI(bundle));
    }

    @Override
    protected void loadUI(Bundle bundle) {
        refreshAttachLayout();

        // update data
        LinkedList<MainItemHolder> n_mainpara;
        if (mBFilter) {
            n_mainpara = new LinkedList<>();
            for (MainItemHolder i : mMainPara) {
                for (String ii : mFilterPara) {
                    if (i.getTag().equals(ii)) {
                        n_mainpara.add(i);
                        break;
                    }
                }
            }
        } else {
            n_mainpara = mMainPara;
        }

        // 设置listview adapter
        YearAdapter mSNAdapter = new YearAdapter(ContextUtil.Companion.getInstance(), n_mainpara);
        mLVShow.setAdapter(mSNAdapter);
        mSNAdapter.notifyDataSetChanged();
    }

    /// BEGIN PRIVATE

    /**
     * 调整数据排序
     */
    private void reorderData() {
        Collections.reverse(mMainPara);
    }

    /**
     * 更新附加layout
     */
    private void refreshAttachLayout() {
        setAttachLayoutVisible(mBFilter || mBSelectSubFilter ? View.VISIBLE : View.GONE);
        setFilterLayoutVisible(mBFilter ? View.VISIBLE : View.GONE);
        setAccpetGiveupLayoutVisible(mBSelectSubFilter ? View.VISIBLE : View.GONE);
    }

    /**
     * 加载详细视图
     * load detail view(for month)
     * @param lv    for view
     * @param tag   tag for data
     */
    private void load_detail_view(ListView lv, String tag) {
        LinkedList<SubItemHolder> llhm = mHMSubPara.get(tag);
        if (!UtilFun.ListIsNullOrEmpty(llhm)) {
            MonthAdapter mAdapter = new MonthAdapter(getContext(), llhm);
            lv.setAdapter(mAdapter);
            mAdapter.notifyDataSetChanged();
            ListViewHelper.INSTANCE.setListViewHeightBasedOnChildren(lv);
        }
    }
    /// END PRIVATE


    /**
     * year item view adapter
     */
    private class YearAdapter extends LVAdapter {
        private static final int SELF_TAG_ID = 0;

        YearAdapter(Context context, List<?> mdata)    {
            super(context, mdata, R.layout.li_yearly_show);
        }

        @Override
        public View getView(final int position, View view, ViewGroup arg2) {
            ViewHolder viewHolder = ViewHolder.get(getRootActivity(), view,
                                        R.layout.li_yearly_show);
            if(null == viewHolder.getSelfTag(SELF_TAG_ID))   {
                viewHolder.setSelfTag(SELF_TAG_ID, new Object());

                final MainItemHolder holder = UtilFun.cast(getItem(position));
                final MainAdapterItem hm = holder.getData();
                final ListView lv = viewHolder.getView(R.id.lv_show_detail);
                final String tag = hm.tag;
                if (EShowFold.getByName(hm.show) == EShowFold.FOLD) {
                    lv.setVisibility(View.GONE);
                } else {
                    lv.setVisibility(View.VISIBLE);
                    if (0 == lv.getCount())
                        load_detail_view(lv, tag);
                }

                View.OnClickListener local_cl = v -> {
                    boolean bf = EShowFold.getByName(hm.show) == EShowFold.FOLD;
                    hm.show = EShowFold.getByFold(!bf).getName();

                    if (bf) {
                        lv.setVisibility(View.VISIBLE);
                        if (0 == lv.getCount())
                            load_detail_view(lv, tag);

                        addUnfoldItem(tag);
                    } else {
                        lv.setVisibility(View.GONE);

                        removeUnfoldItem(tag);
                    }
                };

                // adjust row color
                ConstraintLayout rl = viewHolder.getView(R.id.cl_header);
                rl.setBackgroundColor(0 == position % 2 ?
                        ResourceHelper.INSTANCE.getMCRLVLineOne() : ResourceHelper.INSTANCE.getMCRLVLineTwo());
                rl.setOnClickListener(local_cl);

                // for year
                viewHolder.setText(R.id.tv_year, hm.year);

                // for graph value
                ValueShow vs = viewHolder.getView(R.id.vs_yearly_info);
                HashMap<String, Object> hm_attr = new HashMap<>();
                hm_attr.put(ValueShow.ATTR_PAY_COUNT, hm.yearDetail.mPayCount);
                hm_attr.put(ValueShow.ATTR_PAY_AMOUNT, hm.yearDetail.mPayAmount);
                hm_attr.put(ValueShow.ATTR_INCOME_COUNT, hm.yearDetail.mIncomeCount);
                hm_attr.put(ValueShow.ATTR_INCOME_AMOUNT, hm.yearDetail.mIncomeAmount);
                vs.adjustAttribute(hm_attr);
            }
            return viewHolder.getConvertView();
        }
    }


    /**
     * month item view adapter
     */
    private class MonthAdapter extends LVAdapter {
        private static final int SELF_TAG_ID = 0;

        MonthAdapter(Context context, List<?> sdata)    {
            super(context, sdata, R.layout.li_yearly_show_detail);
        }

        @Override
        public View getView(final int position, View view, ViewGroup arg2) {
            ViewHolder viewHolder = ViewHolder.get(getRootActivity(),
                    view, R.layout.li_yearly_show_detail);
            if(null == viewHolder.getSelfTag(SELF_TAG_ID))   {
                viewHolder.setSelfTag(SELF_TAG_ID, new Object());

                final SubItemHolder holder = UtilFun.cast(getItem(position));
                final SubAdapterItem hm = holder.getData();
                final String sub_tag = hm.subTag;

                final ImageView ib = viewHolder.getView(R.id.iv_action);
                ib.setBackgroundColor(mLLSubFilter.contains(sub_tag) ?
                        ResourceHelper.INSTANCE.getMCRLVItemSel() : ResourceHelper.INSTANCE.getMCRLVItemTransFull());
                ib.setOnClickListener(v -> {
                    String sub_tag1 = hm.subTag;

                    if (!mLLSubFilter.contains(sub_tag1)) {
                        v.setBackgroundColor(ResourceHelper.INSTANCE.getMCRLVItemSel());

                        mLLSubFilter.add(sub_tag1);
                        mLLSubFilterVW.add(v);

                        if (!mBSelectSubFilter) {
                            mBSelectSubFilter = true;
                            refreshAttachLayout();
                        }
                    } else {
                        v.setBackgroundColor(ResourceHelper.INSTANCE.getMCRLVItemTransFull());

                        mLLSubFilter.remove(sub_tag1);
                        mLLSubFilterVW.remove(v);

                        if (mLLSubFilter.isEmpty()) {
                            mLLSubFilterVW.clear();
                            mBSelectSubFilter = false;
                            refreshAttachLayout();
                        }
                    }
                });

                // for show
                viewHolder.setText(R.id.tv_month, hm.month);

                HelperDayNotesInfo.INSTANCE.fillNoteInfo(viewHolder,
                        hm.monthDetail.mPayCount, hm.monthDetail.mPayAmount,
                        hm.monthDetail.mIncomeCount, hm.monthDetail.mIncomeAmount,
                        hm.amount);
            }
            return viewHolder.getConvertView();
        }
    }
}

File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/ListView/LVYearly.kt
package wxm.KeepAccount.ui.data.show.note.ListView

import android.content.Context
import android.os.Bundle
import android.support.constraint.ConstraintLayout
import android.view.View
import android.widget.ImageView
import android.widget.ListView
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import wxm.KeepAccount.R
import wxm.KeepAccount.event.FilterShow
import wxm.androidutil.improve.let1
import wxm.KeepAccount.improve.toSignalMoneyString
import wxm.KeepAccount.ui.base.Helper.ResourceHelper
import wxm.KeepAccount.ui.data.show.note.base.ValueShow
import wxm.KeepAccount.ui.utility.HelperDayNotesInfo
import wxm.KeepAccount.ui.utility.ListViewHelper
import wxm.KeepAccount.ui.utility.NoteDataHelper
import wxm.KeepAccount.utility.ToolUtil
import wxm.androidutil.ui.moreAdapter.MoreAdapter
import wxm.androidutil.ui.view.EventHelper
import wxm.androidutil.ui.view.ViewDataHolder
import wxm.androidutil.ui.view.ViewHolder
import wxm.androidutil.util.UtilFun
import wxm.androidutil.improve.doJudge
import wxm.androidutil.improve.forObj
import wxm.uilib.IconButton.IconButton
import java.util.*
import kotlin.collections.HashMap

/**
 * 年数据视图辅助类
 * Created by WangXM on 2016/9/10.
 */
class LVYearly : LVBase() {
    // 若为true则数据以时间降序排列
    private var mBTimeDownOrder = true

    private val mYearPara: LinkedList<HashMap<String, YearItemHolder>> = LinkedList()
    private val mHMMonthPara: HashMap<String, LinkedList<MonthItemHolder>> = HashMap()

    // for main listview
    internal data class YearData(val year: String) {
        var show: EShowFold = EShowFold.FOLD

        var yearDetail: RecordDetail? = null
        var amount: String? = null
    }

    internal inner class YearItemHolder(tag: String)
        : ViewDataHolder<String, YearData>(tag) {
        override fun getDataByTag(tag: String): YearData {
            val map = YearData(tag)
            NoteDataHelper.getInfoByYear(tag).let {
                map.yearDetail = RecordDetail(it.payCount.toString(), it.szPayAmount,
                        it.incomeCount.toString(), it.szIncomeAmount)

                map.amount = String.format(Locale.CHINA,
                        if (0 < it.balance.toFloat()) "+ %.02f" else "%.02f", it.balance)
            }

            return map
        }
    }

    // for sub listview
    internal data class MonthData(val tag: String, val subTag: String) {
        var month: String? = null
        var monthDetail: RecordDetail? = null
        var amount: String? = null
    }

    internal inner class MonthItemHolder(tag: String)
        : ViewDataHolder<String, MonthData>(tag) {
        override fun getDataByTag(tag: String): MonthData {
            val map = MonthData(tag.substring(0, 4), tag)
            map.month = tag.substring(5, 7).removePrefix("0")

            NoteDataHelper.getInfoByMonth(tag).let {
                map.monthDetail = RecordDetail(it.payCount.toString(), it.szPayAmount,
                        it.incomeCount.toString(), it.szIncomeAmount)

                map.amount = it.balance.toSignalMoneyString()
            }

            return map
        }
    }

    init {
        mBActionExpand = false
        mAHActs = object : ActionHelper() {
            private lateinit var mIBSort: IconButton
            private lateinit var mIBReport: IconButton
            private lateinit var mIBAdd: IconButton
            private lateinit var mIBDelete: IconButton

            override fun initActs(parentView: View) {
                mIBSort = parentView.findViewById(R.id.ib_sort)
                mIBReport = parentView.findViewById(R.id.ib_report)
                mIBAdd = parentView.findViewById(R.id.ib_add)
                mIBDelete = parentView.findViewById(R.id.ib_delete)

                mIBReport.visibility = View.GONE
                mIBAdd.visibility = View.GONE
                mIBDelete.visibility = View.GONE

                mIBSort.setActIcon(mBTimeDownOrder.doJudge(R.drawable.ic_sort_up_1, R.drawable.ic_sort_down_1))
                mIBSort.setActName(mBTimeDownOrder.doJudge(R.string.cn_sort_up_by_time, R.string.cn_sort_down_by_time))

                EventHelper.setOnClickOperator(parentView,
                        intArrayOf(R.id.ib_sort, R.id.ib_refresh),
                        this::onActClick)
            }

            private fun onActClick(v: View) {
                when (v.id) {
                    R.id.ib_sort -> {
                        mBTimeDownOrder = !mBTimeDownOrder

                        mIBSort.setActIcon(mBTimeDownOrder.doJudge(R.drawable.ic_sort_up_1, R.drawable.ic_sort_down_1))
                        mIBSort.setActName(mBTimeDownOrder.doJudge(R.string.cn_sort_up_by_time, R.string.cn_sort_down_by_time))

                        loadUI(null)
                    }

                    R.id.ib_refresh -> {
                        reInitUI()
                    }
                }
            }
        }
    }

    override fun isUseEventBus(): Boolean = true

    /**
     * 过滤视图事件
     *
     * @param event 事件
     */
    @Suppress("unused", "UNUSED_PARAMETER")
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onFilterShowEvent(event: FilterShow) {
    }

    /**
     * 'accept' or 'cancel' when [v] click
     */
    private fun onAcceptOrCancelClick(v: View) {
        when (v.id) {
            R.id.bt_accpet -> if (mBSelectSubFilter) {
                if (!UtilFun.ListIsNullOrEmpty(mLLSubFilter)) {
                    rootActivity!!.jumpByTabName(NoteDataHelper.TAB_TITLE_MONTHLY)
                    EventBus.getDefault().post(
                            FilterShow(NoteDataHelper.TAB_TITLE_YEARLY,
                                    ArrayList<String>(mLLSubFilter)))

                    mLLSubFilter.clear()
                }

                mLLSubFilterVW.forEach {
                    it.isSelected = false
                    it.background.alpha = 0
                }
                mLLSubFilterVW.clear()

                mBSelectSubFilter = false
                refreshAttachLayout()
            }

            R.id.bt_cancel -> {
                mBSelectSubFilter = false
                mLLSubFilter.clear()

                mLLSubFilterVW.forEach {
                    it.isSelected = false
                    it.background.alpha = 0
                }
                mLLSubFilterVW.clear()

                refreshAttachLayout()
            }
        }
    }

    /**
     * 重新加载数据
     */
    override fun initUI(bundle: Bundle?) {
        super.initUI(bundle)
        EventHelper.setOnClickOperator(view!!,
                intArrayOf(R.id.bt_accpet, R.id.bt_cancel),
                this::onAcceptOrCancelClick)

        ToolUtil.runInBackground(activity!!,
                {
                    mYearPara.clear()
                    mHMMonthPara.clear()

                    // for year
                    NoteDataHelper.notesYears.map {
                        HashMap<String, YearItemHolder>()
                                .apply { put(KEY_DATA, YearItemHolder(it)) }
                    }.let1 {
                        mYearPara.addAll(it)
                    }

                    // for month
                    NoteDataHelper.notesMonths.forEach {
                        val ky = it.substring(0, 4)
                        mHMMonthPara[ky].forObj(
                                { ls ->
                                    ls.add(MonthItemHolder(it))
                                    Unit
                                },
                                {
                                    mHMMonthPara[ky] = LinkedList<MonthItemHolder>().apply {
                                        add(MonthItemHolder(it))
                                    }
                                    Unit
                                }
                        )
                    }
                },
                { loadUI(bundle) })
    }

    override fun loadUI(bundle: Bundle?) {
        refreshAttachLayout()

        mLVShow.adapter = YearAdapter(context!!,
                mYearPara.filter { !mBFilter || mFilterPara.contains(it[KEY_DATA]!!.tag) }
                        .sortedWith(Comparator { o1, o2 ->
                            mBTimeDownOrder.doJudge(
                                    o1[KEY_DATA]!!.tag.compareTo(o2[KEY_DATA]!!.tag),
                                    o2[KEY_DATA]!!.tag.compareTo(o1[KEY_DATA]!!.tag))
                        }))
    }

    /// BEGIN PRIVATE

    /**
     * 更新附加layout
     */
    private fun refreshAttachLayout() {
        setAttachLayoutVisible((mBFilter || mBSelectSubFilter).doJudge(View.VISIBLE, View.GONE))
        setFilterLayoutVisible(mBFilter.doJudge(View.VISIBLE, View.GONE))
        setAcceptGiveUpLayoutVisible(mBSelectSubFilter.doJudge(View.VISIBLE, View.GONE))
    }

    /**
     * load month view for [lv] with [tag]
     */
    private fun loadMonthDetailView(lv: ListView, tag: String) {
        mHMMonthPara[tag]?.let {
            LinkedList<HashMap<String, MonthItemHolder>>().apply {
                addAll(it.sortedWith(Comparator { o1, o2 ->
                    mBTimeDownOrder.doJudge(o1.tag.compareTo(o2.tag), o2.tag.compareTo(o1.tag))
                }).map { HashMap<String, MonthItemHolder>().apply { put(KEY_DATA, it) } })
            }.let {
                lv.adapter = MonthAdapter(context!!, it)
            }

            ListViewHelper.setListViewHeightBasedOnChildren(lv)
        }
    }
    /// END PRIVATE

    /**
     * year item view adapter
     */
    private inner class YearAdapter internal constructor(context: Context, mdata: List<Map<String, YearItemHolder>>)
        : MoreAdapter(context, mdata, R.layout.li_yearly_show) {
        override fun loadView(pos: Int, vhHolder: ViewHolder) {
            @Suppress("UNCHECKED_CAST")
            val hm = (getItem(pos) as Map<String, YearItemHolder>)[KEY_DATA]!!.data
            val lv = vhHolder.getView<ListView>(R.id.lv_show_detail)
            val doFold = { lvMonth: ListView, tagData: YearData ->
                lvMonth.visibility = if (tagData.show.isFold()) {
                    View.GONE
                } else {
                    if (0 == lv.count)
                        loadMonthDetailView(lvMonth, tagData.year)

                    View.VISIBLE
                }
            }
            doFold(lv, hm)

            // adjust row color
            vhHolder.getView<ConstraintLayout>(R.id.cl_header).let {
                it.setBackgroundColor((0 == pos % 2)
                        .doJudge(ResourceHelper.mCRLVLineOne, ResourceHelper.mCRLVLineTwo))

                it.setOnClickListener({ _ ->
                    hm.show = hm.show.isFold().doJudge(EShowFold.UNFOLD, EShowFold.FOLD)
                    doFold(lv, hm)
                })
            }

            // for year
            vhHolder.setText(R.id.tv_year, "${hm.year}年")

            // for graph value
            hm.yearDetail!!.let {
                vhHolder.getView<ValueShow>(R.id.vs_yearly_info).adjustAttribute(
                        HashMap<String, Any>().apply {
                            put(ValueShow.ATTR_PAY_COUNT, it.mPayCount)
                            put(ValueShow.ATTR_PAY_AMOUNT, it.mPayAmount)
                            put(ValueShow.ATTR_INCOME_COUNT, it.mIncomeCount)
                            put(ValueShow.ATTR_INCOME_AMOUNT, it.mIncomeAmount)
                        })
            }
        }
    }


    /**
     * month item view adapter
     */
    private inner class MonthAdapter
    internal constructor(context: Context, sdata: List<Map<String, MonthItemHolder>>)
        : MoreAdapter(context, sdata, R.layout.li_yearly_show_detail) {
        override fun loadView(pos: Int, vhHolder: ViewHolder) {
            @Suppress("UNCHECKED_CAST")
            val hm = (getItem(pos) as Map<String, MonthItemHolder>)[KEY_DATA]!!.data
            vhHolder.getView<ImageView>(R.id.iv_action).let {
                it.setBackgroundColor(mLLSubFilter.contains(hm.subTag)
                        .doJudge(ResourceHelper.mCRLVItemSel, ResourceHelper.mCRLVItemTransFull))
                it.setOnClickListener { v ->
                    val subTag = hm.subTag
                    mLLSubFilter.contains(subTag).doJudge(
                            {
                                v.setBackgroundColor(ResourceHelper.mCRLVItemTransFull)

                                mLLSubFilter.remove(subTag)
                                mLLSubFilterVW.remove(v)

                                if (mLLSubFilter.isEmpty()) {
                                    mLLSubFilterVW.clear()
                                    mBSelectSubFilter = false
                                    refreshAttachLayout()
                                }
                            },
                            {
                                v.setBackgroundColor(ResourceHelper.mCRLVItemSel)

                                mLLSubFilter.add(subTag)
                                mLLSubFilterVW.add(v)

                                if (!mBSelectSubFilter) {
                                    mBSelectSubFilter = true
                                    refreshAttachLayout()
                                }
                            }
                    )
                }
            }

            // for show
            vhHolder.setText(R.id.tv_month, hm.month)

            hm.monthDetail?.let {
                HelperDayNotesInfo.fillNoteInfo(vhHolder,
                        it.mPayCount, it.mPayAmount, it.mIncomeCount, it.mIncomeAmount,
                        hm.amount!!)
            }
        }
    }

    companion object {
        private const val KEY_DATA = "data"
    }
}
